src/tree/treemath.rs

```diff
@@ -79,80 +89,104 @@ pub(crate) fn parent_step(x: usize) -> usize {
     (x | (1 << k)) ^ (b << (k + 1))
 }
 
-pub(crate) fn parent(index: NodeIndex, size: LeafIndex) -> NodeIndex {
+pub(crate) fn parent(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
     let x = index.as_usize();
     let n = size.as_usize();
     if index == root(size) {
-        return index;
+        return Err(TreeMathError::RootHasNoParent);
     }
     let mut p = parent_step(x);
     while p >= node_width(n) {
         p = parent_step(p)
     }
-    NodeIndex::from(p)
+    Ok(NodeIndex::from(p))
 }
 
-pub(crate) fn sibling(index: NodeIndex, size: LeafIndex) -> NodeIndex {
-    let p = parent(index, size);
+pub(crate) fn sibling(index: NodeIndex, size: LeafIndex) -> Result<NodeIndex, TreeMathError> {
+    let p = parent(index, size)?;
     match index.cmp(&p) {
         Ordering::Less => right(p, size),
         Ordering::Greater => left(p),
-        Ordering::Equal => p,
+        Ordering::Equal => left(p),
     }
 }
 
 // Ordered from leaf to root
 // Includes neither leaf nor root
-pub(crate) fn dirpath(index: NodeIndex, size: LeafIndex) -> Vec<NodeIndex> {
-    let mut d = vec![];
-    let mut p = parent(index, size);
+pub(crate) fn dirpath(index: NodeIndex, size: LeafIndex) -> Result<Vec<NodeIndex>, TreeMathError> {
     let r = root(size);
-    while p != r {
-        d.push(p);
-        p = parent(p, size);
+    if index == r {
+        return Ok(vec![]);
     }
-    d
+
+    let mut d = vec![];
+    let mut x = index;
+    x = parent(x, size)?;
+    while x != r {
+        d.push(x);
+        x = parent(x, size)?;
+    }
+    return Ok(d);
 }
 
 // Ordered from leaf to root
 // Includes leaf and root
-pub(crate) fn dirpath_long(index: NodeIndex, size: LeafIndex) -> Vec<NodeIndex> {
-    let mut d = vec![index];
-    let mut p = parent(index, size);
+pub(crate) fn dirpath_long(
+    index: NodeIndex,
+    size: LeafIndex,
+) -> Result<Vec<NodeIndex>, TreeMathError> {
     let r = root(size);
     if index == r {
-        return vec![p];
+        return Ok(vec![]);
     }
-    while p != r {
-        d.push(p);
-        p = parent(p, size);
+
+    let mut x = index;
```

Comment:
`x` again isn't muted I think. (same in the other functions)

Reply:
It gets assigned a new value in line 145. Rust complains if I don't assign it as `mut`.

Reply:
Ugh, of course. Sorry 😬